### PR TITLE
✨ implement CAPI IPAM contract support

### LINKED ISSUE #671: Add Support for CAPI IPAM Contract
**User Story**

As a operator I would like to use the new CAPI IPAM contract with metal3 in order to integrate with different IPAM solutions.

**Detailed Description**

The CAPI IPAM contract is now [implemented](https://github.com/kubernetes-sigs/cluster-api/pull/6313) and released as part of CAPI 1.2.0-rc.0.

In order to support it, we'll need to be able to reference IP Pools implemented by various providers. Looking at the current API, I think we can easily do so by extending [`FromPool`](https://github.com/metal3-io/cluster-api-provider-metal3/blob/main/api/v1beta1/metal3datatemplate_types.go#L128-L134) in the DataTemplate with `apiGroup` and `kind` parameters. The new fields should either default to the metal3-ip-address-manager types, or just stay empty. A new reference would then look like this:
```yaml
- key: "something"
  name: "some-pool"
  apiGroup: "ipam.cluster.x-k8s.io"
  kind: "InClusterIPPool"
```
Since only optional fields are added, the change would be fully backwards compatible and doesn't require a new API version.

In code we can then differenciate based on `apiGroup` and `kind`, and either create a metal3-ip-address `IPAddressClaim`, or a CAPI one (ipam.cluster.x-k8s.io).

**Anything else you would like to add:**

I'm happy to work on this.

/kind feature

### PR DESCRIPTION
**Pull request by schrej**
**What this PR does / why we need it**:
Adds support for the CAPI IPAM contract.

To do so, the network related specs of Metal3DataTemplate were adapted to allow referencing other types of Pools than just the one provided by metal3-io/ip-address-manager. Depending on the referenced Pool type, the corresponding controller will create the correct IP claim resource.
For the metadata section, the reference was simply extended with new fields that allow to specify the api group and resource type.
For the network section a new field `FromPoolRef` was added, which can be used as an alternative to "IPAddressFromPool". The new field also supports references to the ip-address-manager Pools.

Since both allocation mechanisms are very similar, the implementation is fairly close as well.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Resolves #671

### CODE REVIEW

**Comment by furkatgofurov7** [comment]
> /assign @furkatgofurov7

👍🏼 I'll take a closer look tomorrow

File: baremetal/metal3data_manager.go [hunk 1]
```diff
@@ -571,6 +690,10 @@ func (m *DataManager) ensureM3IPClaim(ctx context.Context, poolRef corev1.TypedL
 
 // addressFromClaim retrieves the IP address for a ip claim.
```
Comment (furkatgofurov7) [inline baremetal/metal3data_manager.go:691]: nit: addressFromM3Claim retrieves the Metal3IPAddress for a Metal3IPClaim.

**Review by furkatgofurov7 — COMMENTED** [review]
Thanks for this work @schrej, mostly nits for the first iteration. 
I can understand why this PR keeps the allocation logic of Metal3 IPAM as "default" and provides CAPI way as an alternative since it paves the way for smoother migration, but since our current CI  is still not adapted to the new way of IPAM support, I wonder if have you already tested it and could share your thoughts on that regard?

**Comment by schrej** [comment]
I did some manual testing with the in-cluster provider and it works well.
I didn't test backwards compatibility yet, but existing e2e tests should be easy enough to adapt to that. I'll also try to work on some e2e tests for the CAPI contract. For the CAPV provider they're still waiting for a new clusterctl version that supports provisioning ipam providers (the in-cluster provider is compatible already).

**Comment by furkatgofurov7** [comment]
> I did some manual testing with the in-cluster provider and it works well.

Great!

> I didn't test backwards compatibility yet, but existing e2e tests should be easy enough to adapt to that. I'll also try to work on some e2e tests for the CAPI contract. For the CAPV provider they're still waiting for a new clusterctl version that supports provisioning ipam providers (the in-cluster provider is compatible already).

Which version of clusterctl specifically? do you mean v1.3,x?

**Comment by schrej** [comment]
> Which version of clusterctl specifically? do you mean v1.3,x?

Yes! Release is planned for Thursday (1.12.), so not really a blocker.

**Review by lentzi90 — COMMENTED** [review]
Nice! I like the unit tests!

One thing I'm wondering though is if this would require an API version change? It should be backwards compatible, but since we are adding new fields I'm not sure if we can continue with v1beta1.

**Comment by furkatgofurov7** [comment]
> Hey @furkatgofurov7, sorry for the late response, I was on PTO the last few days. I'll update this PR as soon as I find time. 

Hi @schrej , thanks for working on this and no worries, just wanted to check in if this is on track and if there is anything we can help!

> I was struggling with e2e tests, it would be great if someone could assist with that! Maybe we can merge this and work on e2e tests in a separate PR.

Anything specific we can help with e2e tests? I am fine to have a follow-up to deal with e2e tests as well.

**Comment by schrej** [comment]
Tests should be good now.

> Anything specific we can help with e2e tests? I am fine to have a follow-up to deal with e2e tests as well.

I think I was mainly struggling with undestanding how they work in general, and then finding a reasonable test to extend with CAPI IPAM. There doesn't seem to be any basic "deploy a cluster and make sure it works" test, or I'm just not seeing it...

**Comment by lentzi90** [comment]
> I think I was mainly struggling with undestanding how they work in general, and then finding a reasonable test to extend with CAPI IPAM. There doesn't seem to be any basic "deploy a cluster and make sure it works" test, or I'm just not seeing it...

This is a good point. It is weird and surprising that the "least" test we do is a `clusterctl move`. I'll add an issue for an explicit "create cluster and run smoke test". Now that I think about it, the integration test that we have should probably be this. We anyway test `clusterctl move` in most of the other e2e tests so why would it need to be in the basic integration test?

**Comment by furkatgofurov7** [comment]
> We anyway test `clusterctl move` in most of the other e2e tests so why would it need to be in the basic integration test?

indeed, but since e2e tests are quite new and when pivoting was introduced we did not have it at all, so we decided to add it as part of integration tests? /cc @kashifest

**Comment by lentzi90** [comment]
> indeed, but since e2e tests are quite new and when pivoting was introduced we did not have them at all, so we decided to add it as part of integration tests? /cc @kashifest

Let's take this discussion in the issue instead #897

**Comment by furkatgofurov7** [comment]
The CI failure does not look like related to the changes in this PR to me, why the infrastructure is set to v1.4.99 in the tests. 
/cc @mboukhalfa have we changed something in e2e recently?

**Comment by schrej** [comment]
Appears to be a flake, it's green now. I did some manual testing for CAPI IPAM and it still works. If this can be merged without e2e tests, I'd open an issue for those after the change is in.

In addition it might be good to add conditions regarding IP Address allocation on m3d, and a condition on m3m (if that's not there) indicating that it is waiting on data, so it's clearer why provisioning is stalling. Might be worth to analyse a bit more broadly and come up with more conditions on all of the resources as well.

**Review by mboukhalfa — APPROVED** [review]
Thanks I am approving the PR just small nit on the way 

/hold you can unhold once you think it is ready to go in

File: baremetal/metal3data_manager.go [hunk 2]
```diff
@@ -311,13 +313,14 @@ func (m *DataManager) ReleaseLeases(ctx context.Context) error {
...
-	prefix     int
-	gateway    ipamv1.IPAddressStr
+	Address    ipamv1.IPAddressStr
+	Prefix     int
+	Gateway    ipamv1.IPAddressStr
 	dnsServers []ipamv1.IPAddressStr
 }
 
 type reconciledClaim struct {
+	claim      *caipamv1.IPAddressClaim
```
[the remark this answers is not in the document]
Comment (schrej) [reply baremetal/metal3data_manager.go:323]: I've named everything like this, with `m3` as a prefix for the "legacy" implementation, and no prefix for the new one. If we change it here, it probably should be changed everywhere.
Imo it's easier to read like this.

**Comment by kashifest** [comment]
@schrej whats the status of CAPI IPAM btw?

**Review by lentzi90 — COMMENTED** [review]
Thanks for this, it looks solid to me!
/lgtm

**Comment by schrej** [comment]
I think that one is still flaking
/test-centos-e2e-integration-main

**Comment by mboukhalfa** [comment]
> One provider is available, which is very similar to metal3-io/ip-address-management and provides in-cluster IPAM: [telekom/cluster-api-ipam-provider-in-cluster](https://github.com/telekom/cluster-api-ipam-provider-in-cluster) We (= VMware) are currently working on moving the repository to kubernetes-sigs.
> 
> I've also started to work on a Infoblox provider. I'm currently trying to figure out how to handle hostnames, since Infoblox can provide DNS as well, which allows to have automatic DNS records for all machines. I'll probably open an issue on CAPI level at some point.
> 
> Upstream the resources are still experimental and receive occasional changes. Right now we're discussing whether Gateway should be optional: [kubernetes-sigs/cluster-api#8536](https://github.com/kubernetes-sigs/cluster-api/issues/8536)

Would be nice to follow this on some issue on capm3 and set some todo plan for integration tests with capi ipam once it is upstream now we only tested with m3ipam right ?

**Pull request merged**